### ISSUE TITLE
Performance regression after PR#26

### ISSUE THREAD

**Issue by xiaomengy** [facebookresearch/moolib#36]
It seems there is a performance regression after https://github.com/facebookresearch/moolib/pull/26. The perf regression happened mainly when the machine is busy.

A simple reproduce code is shown below. I run the code on a learnfair machine, the average tot_time increased from about 4.40ms to 18.50ms. 

```
from typing import Any, Callable, NoReturn, Tuple

import asyncio
import time

import numpy as np

import torch
import torch.nn as nn
import torch.multiprocessing as mp

import moolib

N = 8192
addr = "127.0.0.1:1234"

# device = "cpu"
device = "cuda"

dynamic_batching = True
num_clients = 32


class Server:
    def __init__(self, dynamic_batching: bool = False) -> None:
        self._addr = addr
        self._dynamic_batching = dynamic_batching

        self._server = None
        self._linear = None
        self._device = device

        self._process = None
        self._loop = None

    @property
    def addr(self) -> str:
        return self._addr

    @property
    def dynamic_batching(self) -> bool:
        return self._dynamic_batching

    def start(self) -> None:
        self._process = mp.Process(target=self.run)
        self._process.start()

    def terminate(self) -> None:
        self._process.terminate()

    def run(self) -> NoReturn:
        self._linear = nn.Linear(N, N).to(self._device)
        self._server = moolib.Rpc()
        self._server.set_name("server")
        self._server.set_timeout(60)
        self._server.listen(addr)

        if self._dynamic_batching:
            self._loop = asyncio.get_event_loop()
            que = self._server.define_queue("run_model",
                                            batch_size=128,
                                            dynamic_batching=True)
            self._loop.create_task(self._async_process(que, self._run_model))
            self._loop.run_forever()
        else:
            self._server.define("run_model", self._run_model)
            while True:
                time.sleep(1)

    def _run_model(self, x: torch.Tensor) -> Tuple[torch.Tensor, float]:
        batch_size = x.size(0) if x.dim() > 1 else None
        t0 = time.perf_counter()
        y = self._linear(x.to(self._device)).sum(dim=-1).cpu()
        t1 = time.perf_counter()
        t = t1 - t0
        if batch_size is not None:
            t = (t, ) * batch_size
        return y, t

    async def _async_process(self, que: moolib.Queue,
                             func: Callable[..., Any]) -> NoReturn:
        try:
            while True:
                ret_cb, args, kwargs = await que
                ret = func(*args, **kwargs)
                ret_cb(ret)
        except asyncio.CancelledError:
            pass


class Client:
    def __init__(self, index: int) -> None:
        self._index = index
        self._addr = addr
        self._client = None
        self._process = None

    @property
    def index(self) -> int:
        return self._index

    def start(self) -> None:
        self._process = mp.Process(target=self.run)
        self._process.start()

    def join(self) -> None:
        self._process.join()

    def terminate(self) -> None:
        self._process.terminate()

    def run(self) -> NoReturn:
        self._client = moolib.Rpc()
        self._client.set_name(f"client-{self._index}")
        self._client.set_timeout(60)
        self._client.connect(self._addr)

        x = torch.randn(N)
        num = 10000

        stats1 = []
        stats2 = []
        for _ in range(num):
            t0 = time.perf_counter()
            _, t = self._client.sync("server", "run_model", x)
            t1 = time.perf_counter()
            stats1.append(t1 - t0)
            stats2.append(t)

        mean1 = np.mean(stats1) * 1000.0
        mean2 = np.mean(stats2) * 1000.0
        print(
            f"[Client-{self._index}], tot_time = {mean1}, run_time = {mean2}")


def main() -> None:
    server = Server(dynamic_batching=dynamic_batching)
    server.start()

    clients = []
    for i in range(num_clients):
        client = Client(i)
        clients.append(client)
        client.start()
    for client in clients:
        client.join()

    server.terminate()


if __name__ == "__main__":
    mp.set_start_method("spawn")
    main()
```

**Comment by condnsdmatters** [comment]
Discussions on this issue have taken place offline over the past few weeks.  A more detailed note on what we will attempt with this is forthcoming.

**Comment by tscmoo** [comment]
Fixed by #38
